controller/src/main/java/io/pravega/controller/store/stream/records/SubscriberSet.java

```diff
@@ -37,37 +37,55 @@
     public static final SubscriberSetSerializer SERIALIZER = new SubscriberSetSerializer();
 
     @Getter
-    private final ImmutableList<String> subscribers;
+    private final ImmutableMap<String, Long> subscribers;
 
     @Builder
-    public SubscriberSet(@NonNull ImmutableList<String> subscribers) {
+    public SubscriberSet(@NonNull ImmutableMap<String, Long> subscribers) {
         this.subscribers = subscribers;
     }
 
     /**
      * This method adds a subscriber in the subscriberSet.
      * @param subscriberSet Subscriber Set.
      * @param subscriber subscriber to be added.
+     * @param generation subscriber generation.
      * @return updated Subscriber Set.
      */
-    public static SubscriberSet add(@NonNull SubscriberSet subscriberSet, @NonNull String subscriber) {
-            ImmutableList.Builder<String> builder = ImmutableList.builder();
-            builder.addAll(subscriberSet.subscribers);
-            builder.add(subscriber);
+    public static SubscriberSet add(@NonNull SubscriberSet subscriberSet, @NonNull String subscriber, long generation) {
+            ImmutableMap.Builder<String, Long> builder = ImmutableMap.builder();
+            builder.putAll(subscriberSet.subscribers);
+            builder.put(subscriber, generation);
             return new SubscriberSet(builder.build());
     }
 
+    /**
+     * This method updates the generation of a subscriber in the subscriberSet.
+     * @param subscriberSet Subscriber Set.
+     * @param subscriber subscriber to be added.
+     * @param generation subscriber generation.
+     * @return updated Subscriber Set.
+     */
+    public static SubscriberSet update(@NonNull SubscriberSet subscriberSet, @NonNull String subscriber, long generation) {
+        ImmutableMap.Builder<String, Long> builder = ImmutableMap.builder();
+        Map<String, Long> otherSubscribers = subscriberSet.getSubscribers().entrySet().stream().filter(e -> !e.getKey().equals(subscriber))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+        builder.putAll(otherSubscribers);
+        builder.put(subscriber, generation);
+        return new SubscriberSet(builder.build());
+    }
+
     /**
      * This method removes a subscriber from the subscriberSet.
      * @param subscriberSet Subscriber Set.
      * @param subscriber subscriber to be removed.
      * @return updated Subscriber Set.
      */
     public static SubscriberSet remove(@NonNull SubscriberSet subscriberSet, @NonNull String subscriber) {
-        ImmutableList.Builder<String> builder = ImmutableList.builder();
-        List<String> otherSubscribers = subscriberSet.getSubscribers().stream().filter(s -> !s.equals(subscriber)).collect(Collectors.toList());
-        builder.addAll(otherSubscribers);
-        return new SubscriberSet(builder.build());
+       ImmutableMap.Builder<String, Long> builder = ImmutableMap.builder();
```

Comment:
same here. we can put all and then call remove on the builder.

Reply:
putAll and Remove is still 2 passes. 
I'm not sure if this would matter as much..... as I'd expect the number of subscribers per Stream to be < 20

Reply:
fixed.